### PR TITLE
SCMI-Server: Add support for multiple source files in build

### PR DESCRIPTION
**Pull request by nicola-mazzucato-arm**
This PR adds support to include c files, other than the "mod_" matching name from the SCP-firmware source tree to build the SCMI Server.
This is required to properly integrate the following change in SCP-firmware:
https://github.com/ARM-software/SCP-firmware/pull/812

### CODE REVIEW

**Comment by nicola-mazzucato-arm** [comment]
@jforissier can you please provide some feedback on this when you have time? Thanks in advance.

**Comment by jforissier** [comment]
@nicola-mazzucato-arm yes I will, but I would also like @etienne-lms to check this when he's back from vacations :wink: Thanks!

**Comment by nicola-mazzucato-arm** [comment]
@jforissier Thanks. Can you please help me figuring out where the failures happened?
The multi-platform build seems to fail on stm32, though it builds just fine locally. Am I missing something?
The other two failures, I am not quite sure where to look at..
Thanks

**Comment by jforissier** [comment]
> @jforissier Thanks. Can you please help me figuring out where the failures happened? The multi-platform build seems to fail on stm32, though it builds just fine locally. Am I missing something?

The CI jobs pulls the SCMI code from the master branch of https://github.com/ARM-software/SCP-firmware.git (see https://github.com/OP-TEE/optee_os/actions/runs/5643523525/job/15392702318?pr=6190#step:5:193), which indeed doesn't have the file `module/scmi_perf/src/scmi_perf_protocol_ops.c`.

> The other two failures, I am not quite sure where to look at.. Thanks

They are unrelated to this PR. They are about insufficient disk space, and I have submitted a fix (https://github.com/OP-TEE/build/pull/671). Sorry about that :-/

**Comment by nicola-mazzucato-arm** [comment]
> The CI jobs pulls the SCMI code from the master branch of https://github.com/ARM-software/SCP-firmware.git (see https://github.com/OP-TEE/optee_os/actions/runs/5643523525/job/15392702318?pr=6190#step:5:193), which indeed doesn't have the file module/scmi_perf/src/scmi_perf_protocol_ops.c.
Right! That seems a bit to create a circular dependency :)
Do you have a way to manually kick off the above tests by fetching the SCP-firmware source from the PR [https://github.com/ARM-software/SCP-firmware/pull/812] to validate the builds? If so, then someone may need to force the merge of this change, but that's up to the maintainers I guess.

> They are unrelated to this PR. They are about insufficient disk space, and I have submitted a fix (https://github.com/OP-TEE/build/pull/671). Sorry about that :-/
Thanks! That's a relief :)

**Comment by jforissier** [comment]
> Do you have a way to manually kick off the above tests by fetching the SCP-firmware source from the PR [https://github.com/[ARM-software/SCP-firmware/pull/812](https://github.com/ARM-software/SCP-firmware/pull/812)] to validate the builds?

You can do it by adding a commit that changes the CI file (`.github/workflows/ci.yml`). Replace `function download_scp_firmware() { ... }` with the following:
```
function download_scp_firmware() { git clone --single-branch https://github.com/ARM-software/SCP-firmware.git $HOME/scp-firmware && { cd $HOME/scp-firmware && git fetch origin pull/812/head && git checkout FETCH_HEAD; } || echo Nevermind; }
```
Call this commit "[TESTING] Update CI to test with SCP firmware PR 812" for instance.

**Comment by jforissier** [comment]
@nicola-mazzucato-arm I'm sorry the line should be:
```
function download_scp_firmware() { git clone --single-branch https://github.com/ARM-software/SCP-firmware.git $HOME/scp-firmware && ( cd $HOME/scp-firmware && git fetch origin pull/812/head && git checkout FETCH_HEAD; ) || echo Nevermind; }
```
...that is, with parentheses `( )` instead of curly braces `{ }`, so that a sub-shell is forked and the `cd` command only has local scope.

**Comment by nicola-mazzucato-arm** [comment]
> @nicola-mazzucato-arm I'm sorry the line should be:
> 
> ```
> function download_scp_firmware() { git clone --single-branch https://github.com/ARM-software/SCP-firmware.git $HOME/scp-firmware && ( cd $HOME/scp-firmware && git fetch origin pull/812/head && git checkout FETCH_HEAD; ) || echo Nevermind; }
> ```
> 
> ...that is, with parentheses `( )` instead of curly braces `{ }`, so that a sub-shell is forked and the `cd` command only has local scope.

Thanks @jforissier , I was just adding the ```cd ..```. I'll try your suggestion anyway :)

File: core/lib/scmi-server/sub.mk [hunk 1]
```diff
@@ -148,6 +151,20 @@ ifneq (,$$(wildcard $(scpfw-path)/$3/$2/include/*))
...
 srcs-$(CFG_SCPFW_MOD_$4) += $(scpfw-path)/$3/$2/src/mod_$1.c
+
+# SCMI_Perf has components that can be added conditionally
+ifeq ($(1), scmi_perf)
+
+ifneq (,$(filter y, $(CFG_SCPFW_SCMI_PERF_PROTOCOL_OPS)))
+srcs-$(CFG_SCPFW_MOD_SCMI_PERF) += $(scpfw-path)/$3/$2/src/scmi_perf_protocol_ops.c
+endif
+
+ifneq (,$(filter y, $(CFG_SCPFW_SCMI_PERF_FAST_CHANNELS)))
```
Comment (etienne-lms) [inline core/lib/scmi-server/sub.mk:162]: Do you foresee added CFG_ dependencies? Unless what, `ifneq ($(CFG_SCPFW_SCMI_PERF_xxx),y)` is simpler.
Ditto at line 159.

Comment (nicola-mazzucato-arm) [reply core/lib/scmi-server/sub.mk:162]: Not really.. I will change this and push an updated PR. Thanks!

**Comment by nicola-mazzucato-arm** [comment]
FYI @jforissier @etienne-lms I have updated the PR. Thanks

**Comment by nicola-mazzucato-arm** [comment]
@jforissier can you kindly kick-off all the tests? Thanks in advance :)

**Comment by jforissier** [comment]
> @jforissier can you kindly kick-off all the tests? Thanks in advance :)

Done.

Can we expect the SCP-firmware to be merged before merging this? I do not want to merge something that breaks CI for everyone.

**Comment by nicola-mazzucato-arm** [comment]
> > @jforissier can you kindly kick-off all the tests? Thanks in advance :)
> 
> Done.
> 
> Can we expect the SCP-firmware to be merged before merging this? I do not want to merge something that breaks CI for everyone.

Thanks! I was hoping the other way round :). The moment you merge this PR I will merge PR812 in SCP. That should align the two repos and CI should be happy in either sides.

**Comment by jforissier** [comment]
Then the SCP build has to be disabled first in CI. Add a commit (first commit in the series) that explains that the build is going to be broken by subsequent commits and will require 812 to be merged in SCP-firmware and therefore the build is disabled temporarily. Then when 812 is merged you can simply revert that commit.

**Comment by nicola-mazzucato-arm** [comment]
> Then the SCP build has to be disabled first in CI. Add a commit (first commit in the series) that explains that the build is going to be broken by subsequent commits and will require 812 to be merged in SCP-firmware and therefore the build is disabled temporarily. Then when 812 is merged you can simply revert that commit.

Thanks @jforissier for the suggestion. PR has been resubmitted with the scmi-server temporarily disabled.

**Comment by jforissier** [comment]
One typo: s/rever/revert
I suspect checkpactch will complain on the long line with the URL, please use this format instead:
```
Link: https://some-url/ [1]
Signed-off-by: ...
```
(No need for a blank line between the two)
One last thing, it seems easier to just not download the source code (make the download function do nothing) instead of changing 3 lines to `=n`, but I am OK with your approach too.
For this commit:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by nicola-mazzucato-arm** [comment]
> One typo: s/rever/revert I suspect checkpactch will complain on the long line with the URL, please use this format instead:
> 
> ```
> Link: https://some-url/ [1]
> Signed-off-by: ...
> ```
> 
> (No need for a blank line between the two) One last thing, it seems easier to just not download the source code (make the download function do nothing) instead of changing 3 lines to `=n`, but I am OK with your approach too. For this commit:
> 
> `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Thank you @jforissier 

For this new force-push:
- fixed typo in commit msg
- emptied out `download_scp_firmware()` instead of setting `CFG_SCMI_SCPFW` to `n` as suggested
Thanks!

**Comment by nicola-mazzucato-arm** [comment]
@jforissier any clue why the multi-platform failed? Thanks

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -81,7 +81,7 @@ jobs:
 
           function _make() { make -j$(nproc) -s O=out $*; }
           function download_plug_and_trust() { mkdir -p $HOME/se050 && git clone --single-branch -b v0.4.2 https://github.com/foundriesio/plug-and-trust $HOME/se050/plug-and-trust || (rm -rf $HOME/se050 ; echo Nervermind); }
-          function download_scp_firmware() { git clone --single-branch https://github.com/ARM-software/SCP-firmware.git $HOME/scp-firmware || echo Nervermind; }
+          function download_scp_firmware() { #Temporarily disabled }
```
Comment (jforissier) [inline .github/workflows/ci.yml:84]: You cannot use a comment here, it comments out the closing brace. Add it on the previous line.

Comment (nicola-mazzucato-arm) [reply .github/workflows/ci.yml]: > You cannot use a comment here, it comments out the closing brace. Add it on the previous line.

Got it, thanks!
New set submitted.

**Comment by nicola-mazzucato-arm** [comment]
@jforissier anything else required on my end for this to be merged? Thanks

**Comment by jforissier** [comment]
> @jforissier anything else required on my end for this to be merged? Thanks

No, we're good, thanks!

**Pull request merged**